include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -104,34 +118,41 @@ struct GGX<T,false NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
         return scalar_type(1.0) / (absNdotX + devsh_part);
     }
 
-    template<class LS, class Interaction NBL_FUNC_REQUIRES(LightSample<LS> && surface_interactions::Isotropic<Interaction>)
-    scalar_type correlated_wo_numerator(NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction, BxDFClampMode _clamp)
+    // without numerator, numerator is 2 * NdotV * NdotL, we factor out 4 * NdotV * NdotL, hence 0.5
+    template<class Query, class LS, class Interaction NBL_FUNC_REQUIRES(ggx_concepts::G2XQuery<Query> && LightSample<LS> && surface_interactions::Isotropic<Interaction>)
+    scalar_type correlated(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction)
     {
-        // numerator is 2 * NdotV * NdotL, we factor out 4 * NdotV * NdotL, hence 0.5
-        scalar_type Vterm = _sample.getNdotL(_clamp) * devsh_part(interaction.getNdotV2());
-        scalar_type Lterm = interaction.getNdotV(_clamp) * devsh_part(_sample.getNdotL2());
+        BxDFClampMode _clamp = query.getClampMode();
```

Comment:
imho the clamp mode should probably be a `NBL_STATIC_CONSTEXPR_INLINE` in the query struct